File: https:/github.com/abhinavtripathi17/DSA-Leetcode.git/300-LongestIncreasingSubsequence/300-LongestIncreasingSubsequence.java
// Last updated: 8/4/2026, 7:17:29 PM
class Solution {
    public int lengthOfLIS(int[] nums) {
        int n = nums.length;
        List<Integer>ll = new ArrayList<>();

        int cnt = 1;
        ll.add(nums[0]);

        for(int i = 1 ; i < n ; i++){
            if(nums[i] > ll.get(ll.size() - 1)){
                ll.add(nums[i]);
                cnt++;
            }
            else{
                int idx = Collections.binarySearch(ll , nums[i]);

                if(idx < 0){
                    idx = -idx - 1;
                }
                ll.set(idx , nums[i]);
            }
        }
        return cnt;
    } 
    
}

File: https:/github.com/abhinavtripathi17/DSA-Leetcode.git/300-LongestIncreasingSubsequence/300-LongestIncreasingSubsequence.cpp
// Last updated: 8/4/2026, 7:17:23 PM
class Solution {
public:
    
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>>dp(n , vector<int>(n+1 , -1));
        return lis(nums , 0 , -1, n , dp);
    }

    int lis(vector<int>& nums , int idx , int prevIdx , int n , vector<vector<int>>& dp){

        if(idx == n){
            return 0;
        }

        if(dp[idx][prevIdx+1] != -1) return dp[idx][prevIdx+1];
        int len = lis(nums , idx+1 , prevIdx , n , dp);

        if(prevIdx == -1 || nums[idx] > nums[prevIdx]){
            len = max(len , 1 + lis(nums , idx+1 , idx , n , dp));
        }

        return dp[idx][prevIdx+1] = len;
    }
};